source/validate_types.cpp

```diff
@@ -38,85 +38,162 @@ using std::find;
 using std::string;
 using std::unordered_set;
 using std::vector;
+using namespace libspirv;
 
 namespace {
-const vector<vector<SpvOp>>& GetModuleOrder() {
+bool IsInstructionInLayoutSection(ModuleLayoutSection layout, SpvOp op) {
   // See Section 2.4
+  bool out = false;
   // clang-format off
-  static const vector<vector<SpvOp>> moduleOrder = {
-    {SpvOpCapability},
-    {SpvOpExtension},
-    {SpvOpExtInstImport},
-    {SpvOpMemoryModel},
-    {SpvOpEntryPoint},
-    {SpvOpExecutionMode},
-    {
-      // first set of debug instructions
-      SpvOpSourceContinued,
-      SpvOpSource,
-      SpvOpSourceExtension,
-      SpvOpString,
-    },
-    {
-      // second set of debug instructions
-      SpvOpName,
-      SpvOpMemberName
-    },
-    {
-      // annotation instructions
-      SpvOpDecorate,
-      SpvOpMemberDecorate,
-      SpvOpGroupDecorate,
-      SpvOpGroupMemberDecorate,
-      SpvOpDecorationGroup
-    },
-    {
-      // All type and constant instructions
-      SpvOpTypeVoid,
-      SpvOpTypeBool,
-      SpvOpTypeInt,
-      SpvOpTypeFloat,
-      SpvOpTypeVector,
-      SpvOpTypeMatrix,
-      SpvOpTypeImage,
-      SpvOpTypeSampler,
-      SpvOpTypeSampledImage,
-      SpvOpTypeArray,
-      SpvOpTypeRuntimeArray,
-      SpvOpTypeStruct,
-      SpvOpTypeOpaque,
-      SpvOpTypePointer,
-      SpvOpTypeFunction,
-      SpvOpTypeEvent,
-      SpvOpTypeDeviceEvent,
-      SpvOpTypeReserveId,
-      SpvOpTypeQueue,
-      SpvOpTypePipe,
-      SpvOpTypeForwardPointer,
-      SpvOpConstantTrue,
-      SpvOpConstantFalse,
-      SpvOpConstant,
-      SpvOpConstantComposite,
-      SpvOpConstantSampler,
-      SpvOpConstantNull,
-      SpvOpSpecConstantTrue,
-      SpvOpSpecConstantFalse,
-      SpvOpSpecConstant,
-      SpvOpSpecConstantComposite,
-      SpvOpSpecConstantOp,
-      SpvOpVariable,
-      SpvOpLine
-    }
-  };
+  switch (layout) {
+    case kLayoutCapabilities:  out = op == SpvOpCapability;    break;
```

Comment:
I tend to just do an early return in situations like this.  (Also, that's not forbidden by Google style guide, but it is by some.   You would also tend to add a dummy "return <something>" at the end of the function to avoid warnings in some compilers.)

Coder's choice.  (i.e. ok to keep this)